projects/driver_controls/test/test_throttle.c

```diff
@@ -0,0 +1,162 @@
+      data->zone_thresholds[channel][zone][thresh] = threshes[zone][thresh];
+    }
+  }
+}
+
+void setup_test(void) {
+  gpio_init();
+  interrupt_init();
+  gpio_it_init();
+  soft_timer_init();
+  I2CSettings i2c_settings = {
+    .speed = I2C_SPEED_FAST,                    //
+    .scl = { .port = GPIO_PORT_B, .pin = 10 },  //
+    .sda = { .port = GPIO_PORT_B, .pin = 11 },  //
+  };
+  i2c_init(TEST_ADS1015_I2C_PORT, &i2c_settings);
+  GPIOAddress ready_pin = {
+    .port = GPIO_PORT_B,  //
+    .pin = 2,             //
+  };
+  ads1015_init(&ads1015_storage, TEST_ADS1015_I2C_PORT, TEST_ADS1015_ADDR, &ready_pin);
+  prv_set_calibration_data(THROTTLE_CHANNEL_MAIN, s_threshes_main, &calibration_data);
+  prv_set_calibration_data(THROTTLE_CHANNEL_SECONDARY, s_threshes_secondary, &calibration_data);
+  prv_set_calibration_data_tolerance(s_tolerance, &calibration_data);
+}
+
+void teardown_test(void) {}
+
+void test_throttle_init_invalid_args(void) {
+  // Test with valid arguments.
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK,
+                    throttle_init(&throttle_storage, &calibration_data, &ads1015_storage,
+                                  ADS1015_CHANNEL_0, ADS1015_CHANNEL_1));
+  // Check for null pointers.
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    throttle_init(NULL, &calibration_data, &ads1015_storage, ADS1015_CHANNEL_0,
+                                  ADS1015_CHANNEL_1));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    throttle_init(&throttle_storage, NULL, &ads1015_storage, ADS1015_CHANNEL_0,
+                                  ADS1015_CHANNEL_1));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    throttle_init(&throttle_storage, &calibration_data, NULL, ADS1015_CHANNEL_0,
+                                  ADS1015_CHANNEL_1));
+  // Check for invalid channels.
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    throttle_init(&throttle_storage, &calibration_data, &ads1015_storage,
+                                  NUM_ADS1015_CHANNELS, ADS1015_CHANNEL_1));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    throttle_init(&throttle_storage, &calibration_data, &ads1015_storage,
+                                  ADS1015_CHANNEL_0, NUM_ADS1015_CHANNELS));
+}
+
+void test_throttle_get_pos_invalid_args(void) {
+  ThrottlePosition position;
+  throttle_init(&throttle_storage, &calibration_data, &ads1015_storage, ADS1015_CHANNEL_0,
+                ADS1015_CHANNEL_1);
+  // Check for null pointers.
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, throttle_get_position(NULL, &position));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, throttle_get_position(&throttle_storage, NULL));
+}
+
+void test_throttle_verify_zone_event(void) {
+  ThrottlePosition position;
+  throttle_init(&throttle_storage, &calibration_data, &ads1015_storage, ADS1015_CHANNEL_0,
+                ADS1015_CHANNEL_1);
+  // Brake zone.
+  s_mocked_reading =
```

Comment:
I think it would make more sense to be able to set the mocked reading separately so you can test when they diverge. (failure conditions)

Reply:
Could you rephrase please? Not sure what you mean by separate.